NDTensors/src/lib/SparseArrayDOKs/src/sparsearraydok.jl

```diff
@@ -1,10 +1,11 @@
 using Accessors: @set
 using Dictionaries: Dictionary, set!
+using MacroTools: @capture
 using ..SparseArrayInterface:
   SparseArrayInterface, AbstractSparseArray, getindex_zero_function
 
 # TODO: Parametrize by `data`?
-struct SparseArrayDOK{T,N,Zero} <: AbstractSparseArray{T,N}
+mutable struct SparseArrayDOK{T,N,Zero} <: AbstractSparseArray{T,N}
```

Comment:
What I don't like about marking the entire struct as `mutable` is that makes all of the fields mutable, while we only want `dims` to be mutable.

In Julia 1.8 and above we could do the following:
```julia
mutable struct SparseArrayDOK{T,N,Zero} <: AbstractSparseArray{T,N}
  const data::Dictionary{CartesianIndex{N},T}
  dims::NTuple{N,Int}
  const zero::Zero
end
```
as described in the Julia docs here: https://docs.julialang.org/en/v1/manual/types/#Mutable-Composite-Types but I want to support Julia 1.6 for now se we'll have to hold off on that design.

For now what we can do is the following:
```julia
struct SparseArrayDOK{T,N,Zero} <: AbstractSparseArray{T,N}
  data::Dictionary{CartesianIndex{N},T}
  dims::Ref{NTuple{N,Int}}
  zero::Zero
end
```
i.e. not mark the struct as `mutable` but instead make the `dims` field a reference to the tuple of dimension lengths.

Then, we can define:
```julia
Base.size(a::SparseArrayDOK) = a.dims[]

function Base.resize!(a::SparseArrayDOK{<:Any,N}, new_size::NTuple{N,Integer}) where {N}
  a.dims[] = new_size
  return a
end
```
etc. Some of the constructors would of course need to be updated, however I think we could just do that with an inner constructor:
```julia
struct SparseArrayDOK{T,N,Zero} <: AbstractSparseArray{T,N}
  data::Dictionary{CartesianIndex{N},T}
  dims::Ref{NTuple{N,Int}}
  zero::Zero
  function SparseArrayDOK{T,N,Zero}(data, dims::NTuple{N,Int}, zero)
    return new{T,N,Zero}(data, Ref(dims), zero)
  end
end
```
and ideally keep the rest of the constructors the same.

Reply:
Yes, I sort of cringed having to put `mutable` on the type, so this a nice compromise. Good to know about the better option post Julia 1.8. I just made the above changes.